lxd/instance_efi_vars.go

```diff
@@ -0,0 +1,218 @@
+		return response.SmartError(err)
+	}
+
+	if inst.Type() != instancetype.VM {
+		return response.BadRequest(fmt.Errorf("EFI variables manipulation supported for VM type instances only"))
+	}
+
+	instanceEFI, err := inst.EFIVarsRead()
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	etag := []any{inst.Architecture(), inst.LocalConfig(), instanceEFI}
+
+	return response.SyncResponseETag(true, instanceEFI, etag)
+}
+
+// swagger:operation PUT /1.0/instances/{name}/efi_vars instances instance_efi_vars_put
+//
+//	Set the VM instance EFI variables
+//
+//	Sets a specific instance EFI variables.
+//
+//	---
+//	consumes:
+//	  - application/json
+//	produces:
+//	  - application/json
+//	parameters:
+//	  - in: query
+//	    name: project
+//	    description: Project name
+//	    type: string
+//	    example: default
+//	  - in: body
+//	    name: instanceEFI
+//	    description: EFI variables update request
+//	    schema:
+//	      $ref: "#/definitions/InstanceEFI"
+//	responses:
+//	  "200":
+//	    $ref: "#/responses/EmptySyncResponse"
+//	  "400":
+//	    $ref: "#/responses/BadRequest"
+//	  "403":
+//	    $ref: "#/responses/Forbidden"
+//	  "404":
+//	    $ref: "#/responses/NotFound"
+//	  "500":
+//	    $ref: "#/responses/InternalServerError"
+func instanceEFIVarsPut(d *Daemon, r *http.Request) response.Response {
+	// Don't mess with instance while in setup mode.
+	<-d.waitReady.Done()
+
+	s := d.State()
+
+	instanceType, err := urlInstanceTypeDetect(r)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	projectName := request.ProjectParam(r)
+
+	// Get the container
+	name, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if shared.IsSnapshot(name) {
+		return response.BadRequest(fmt.Errorf("Invalid instance name"))
+	}
+
+	// Handle requests targeted to a container on a different node
+	resp, err := forwardedResponseIfInstanceIsRemote(s, r, projectName, name, instanceType)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if resp != nil {
+		return resp
+	}
+
+	unlock, err := instanceOperationLock(s.ShutdownCtx, projectName, name)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	defer unlock()
+
+	inst, err := instance.LoadByProjectAndName(s, projectName, name)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if inst.Type() != instancetype.VM {
+		return response.BadRequest(fmt.Errorf("EFI variables manipulation supported for VM type instances only"))
+	}
+
+	instanceEFI, err := inst.EFIVarsRead()
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Validate the ETag
+	etag := []any{inst.Architecture(), inst.LocalConfig(), instanceEFI}
+	err = util.EtagCheck(r, etag)
+	if err != nil {
+		return response.PreconditionFailed(err)
+	}
+
+	configRaw := api.InstanceEFI{}
+	err = json.NewDecoder(r.Body).Decode(&configRaw)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	err = configRaw.CheckValid()
```

Comment:
Can we move this CheckValid call into `EFIVarsUpdate()` so its impossible to write invalid config?

Reply:
Fixed, thanks!